### PR TITLE
PageContainer + dual image resolution loading

### PR DESCRIPTION
**Pull request by cdrini**
For (code) reviewers: Going commit by commit might be best. I tried to leave lots of context in my commit messages :+1:

See it live! https://ia-petabox-drini.archive.org/details/goodytwoshoes00newyiala

https://user-images.githubusercontent.com/6251786/113650258-e8fdbc80-965d-11eb-9950-e687c1e0483f.mp4

https://user-images.githubusercontent.com/6251786/113649551-a1c2fc00-965c-11eb-98da-e983079e67d3.mp4

Todo:

- [x] Fix failing unit tests
- [x] Bug: Front/back cover broken in Mode2Up
- [x] Fix failing E2E tests
- [x] Add more unit tests
- [x] Manual tests + performance tests
    - Did some basic tests and it seems like it's making the same number of requests as before. Didn't test super thoroughly though.
- [x] Toss in a review app for QA https://ia-petabox-drini.archive.org/details/goodytwoshoes00newyiala
- [x] Record some before/after cause they look nice :)

Todo in future PR:
- Limit the cache to powers of 2; that'll result in more cache hits.

### CODE REVIEW

**Review by iisa — APPROVED** [review]
:shipit: 
`PageContainer` + `ImageCache` reducer list per image is everything BookReader has been dreaming of. ☁️ 

the extraction of PageContainer alone is a big upgrade to the core code. Fantastic, drini!! 🙌 

just a few comments, questions

File: src/js/BookReader.js [hunk 1]
```diff
@@ -2219,33 +2171,17 @@ BookReader.prototype.canSwitchToMode = function(mode) {
  * @return {string}
  */
 BookReader.prototype._getPageURISrcset = function(index, reduce, rotate) {
-  if (index < 0 || index >= this._models.book.getNumLeafs()) { // Synthesize page
-    return "";
-  }
+  const page = this._models.book.getPage(index, false);
+  // Synthesize page
+  if (!page) return "";
```
[the remark this answers is not in the document]
Comment (cdrini) [reply src/js/BookReader.js:2176]: It does a `.join()` before returning, so result is always a string :+1:

File: src/js/BookReader/BookModel.js [hunk 2]
```diff
@@ -439,6 +444,31 @@ class PageModel {
     return this.book.getPageURI(this.index, reduce, rotate);
   }
 
+  /**
+   * Returns the srcset with correct URIs or void string if out of range
+   * Also makes the reduce argument optional
+   * @param {number} reduce
+   * @param {number} [rotate]
+   */
+  getURISrcSet(reduce, rotate = 0) {
```
[the remark this answers is not in the document]
Comment (cdrini) [reply src/js/BookReader/BookModel.js:453]: Yep! Copypasta from BookReader.js

File: src/js/BookReader/ModeThumb.js [hunk 3]
```diff
@@ -229,23 +215,9 @@ export class ModeThumb {
     // highlight current page
     this.br.$('.pagediv' + this.br.currentIndex()).addClass('BRpagedivthumb_highlight');
 
-    this.br.lazyLoadThumbnails();
-
     this.br.updateToolbarZoom(this.br.reduce);
   }
 
-  lazyLoadThumbnails() {
```
[the remark this answers is not in the document]
Comment (cdrini) [reply src/js/BookReader/ModeThumb.js:237]: 🪓 🪓 🪓 !

File: src/js/BookReader/PageContainer.js [hunk 4]
```diff
@@ -0,0 +1,75 @@
+// @ts-check
+/** @typedef {import('./BookModel.js').PageModel} PageModel */
+/** @typedef {import('./ImageCache.js').ImageCache} ImageCache */
+
+
+export class PageContainer {
```
[the remark this answers is not in the document]
Comment (cdrini) [reply src/js/BookReader/PageContainer.js:6]: Super pleased with this one :)

File: tests/BookReader/BookModel.test.js [hunk 5]
```diff
@@ -218,6 +218,25 @@ describe('_getDataFlattened', () => {
   });
 });
 
+describe('getPage', () => {
+  test('loops around by default', () => {
+    const bm = new BookModel({ data: SAMPLE_DATA });
+    expect(bm.getPage(-1).index).toBe(3);
+    expect(bm.getPage(-2).index).toBe(2);
+    expect(bm.getPage(-3).index).toBe(1);
```
[the remark this answers is not in the document]
Comment (cdrini) [reply tests/BookReader/BookModel.test.js:226]: _Ye who seeks a page, and is true of `loop`, shall always a page find._

File: tests/BookReader/PageContainer.test.js [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+  test('adds/removes loading indicators while loading', () => {
+    const fakeImageCache = {
+      imageLoaded: () => false,
+      image: () => $('<img/>'),
+      getBestLoadedReduce: () => undefined,
+    };
+    const pc = new PageContainer({index: 12}, {imageCache: fakeImageCache, loadingImage: 'loading.gif'});
+    pc.update({ reduce: 7 });
+    expect(pc.$container.hasClass('BRpageloading')).toBe(true);
+    // See https://github.com/jsdom/jsdom/issues/3169
```
[the remark this answers is not in the document]
Comment (cdrini) [reply tests/BookReader/PageContainer.test.js:82]: #le-sigh

File: tests/e2e/viewmode.test.js [hunk 7]
```diff
@@ -24,7 +24,7 @@ test('Clicking `view mode` cycles through view modes', async t => {
   await t.click(nav.desktop.viewmode);
   const onePageViewContainer = Selector('.BRpageview');
   await t.expect(onePageViewContainer.visible).ok();
-  const onePageImages = onePageViewContainer.find('.BRpagecontainer.BRmode1up');
+  const onePageImages = onePageViewContainer.find('.BRmode1up .BRpagecontainer');
```
[the remark this answers is not in the document]
Comment (cdrini) [reply tests/e2e/viewmode.test.js:27]: So much easier to reason about for me :)

File: tests/e2e/helpers/base.js [hunk 8]
```diff
@@ -182,29 +185,20 @@ export function runBaseTests (br) {
...
 
   test('Clicking `1 page view` brings up 1 at a time', async t => {
     const { nav } = br;
     await t.click(nav.desktop.mode1Up);
-    const onePageViewContainer = Selector('.BRpageview');
-    await t.expect(onePageViewContainer.visible).ok();
-    const images = onePageViewContainer.find('.BRpagecontainer.BRmode1up');
     // we usually pre-fetch the page in question & the 2 after it
-    await t.expect(images.count).gte(3);
+    await t.expect(Selector('.BRpagecontainer').count).gte(1);
```
[the remark this answers is not in the document]
Comment (cdrini) [reply tests/e2e/helpers/base.js:195]: Let me see if it works with `gte(3)` ; I can't recall why I changed this...

[the remark this answers is not in the document]
Comment (cdrini) [reply tests/e2e/helpers/base.js:195]: Works locally! Updating...

Comment (cdrini) [inline tests/e2e/helpers/base.js:195]:
```suggestion
    await t.expect(Selector('.BRpagecontainer').count).gte(3);
```

**Comment by cdrini** [comment]
(Tested on IE11 just to be safe 😅 Should work anywhere/everywhere!)

**Pull request merged**